From f54af2f478e7869634e492ff4c1caf125ee92167 Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 17:32:54 -0400
Subject: [infiniband] qib: Fix QP check when looping back to/from QP1

Message-id: <df20b0c20cd3fd45805f2e5517d758b321ee88e4.1411579814.git.dledford@redhat.com>
Patchwork-id: 96685
O-Subject: [Patch RHEL7 01/19] IB/qib: Fix QP check when looping back to/from QP1
Bugzilla: 1085961
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: Ira Weiny <ira.weiny@intel.com>

The GSI QP type is compatible with and should be allowed to send data
to/from any UD QP.  This was found when testing ibacm on the same node
as an SA.

Cc: <stable@vger.kernel.org>
Reviewed-by: Mike Marciniszyn <mike.marciniszyn@intel.com>
Signed-off-by: Ira Weiny <ira.weiny@intel.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 6e0ea9e6cbcead7fa8c76e3e3b9de4a50c5131c5)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/qib/qib_ud.c b/drivers/infiniband/hw/qib/qib_ud.c
index d6c7fe7..3ad651c 100644
--- a/drivers/infiniband/hw/qib/qib_ud.c
+++ b/drivers/infiniband/hw/qib/qib_ud.c
@@ -57,13 +57,20 @@ static void qib_ud_loopback(struct qib_qp *sqp, struct qib_swqe *swqe)
  struct qib_sge *sge;
  struct ib_wc wc;
  u32 length;
+ enum ib_qp_type sqptype, dqptype;
 
  qp = qib_lookup_qpn(ibp, swqe->wr.wr.ud.remote_qpn);
  if (!qp) {
   ibp->n_pkt_drops++;
   return;
  }
- if (qp->ibqp.qp_type != sqp->ibqp.qp_type ||
+
+ sqptype = sqp->ibqp.qp_type == IB_QPT_GSI ?
+   IB_QPT_UD : sqp->ibqp.qp_type;
+ dqptype = qp->ibqp.qp_type == IB_QPT_GSI ?
+   IB_QPT_UD : qp->ibqp.qp_type;
+
+ if (dqptype != sqptype ||
      !(ib_qib_state_ops[qp->state] & QIB_PROCESS_RECV_OK)) {
   ibp->n_pkt_drops++;
   goto drop;
-- 
1.7.1